common/src/main/java/com/revenuecat/purchases/common/BillingAbstract.kt

```diff
@@ -79,7 +79,8 @@ abstract class BillingAbstract {
         appUserID: String,
         purchasingData: PurchasingData,
         replaceProductInfo: ReplaceProductInfo?,
-        presentedOfferingIdentifier: String?
+        presentedOfferingIdentifier: String?,
+        isPersonalizedPrice: Boolean = false
```

Comment:
Since this is very Google specific, I _think_ I like the idea of calling this `StoreOptions` 

```kotlin
interface StorePurchasingOptions {
}

data class GoogleStorePurchasingOptions(
    val isPersonalizedPrice: Boolean
) : StorePurchasingOptions()

// This may not even be needed
data class AmazonStorePurchasingOptions(

) : StorePurchasingOptions()
```

```kotlin
    override fun makePurchaseAsync(
        activity: Activity,
        appUserID: String,
        purchasingData: PurchasingData,
        replaceProductInfo: ReplaceProductInfo?,
        presentedOfferingIdentifier: String?
        presentedOfferingIdentifier: String?,
        options: StorePurchasingOptions
    )
```

And then cast `StoreOptions` to `GoogleStorePurchasingOptions` inside of `makePurchaseAsync()`. This way Amazon and Google can keep their options independent

This _may_ be a little over engineering for now but just a though 🤷‍♂️

Reply:
i love the idea... trying to think if there could be other purchasingOptions in the future...

my only argument kinda "against" this is that the user still has to set the google-specific property on the public api, so it's almost like wrapping it to immediately unwrap it, without saving the user from having to think about google stuff. idk if i'm making sense, just brain dumping...

Reply:
@joshdholtz i didn't end up going this route because it felt a bit like over-engineering. i could definitely be convinced though, i do like the idea. good news is, it's not a public API so we can always add it at a later date...

Reply:
Hmm so the API for purchases becomes a lot more complex with the extra cardinality of adding a new optional parameter and supporting all the different combinations. This would only get worse if we ever have to add another optional parameter in the future. What do you think about applying Josh's idea to the public API? So the purchase methods in purchases would look like:
```
fun purchaseProduct(
    activity: Activity,
    storeProduct: StoreProduct
    listener: ProductChangeCallback
)
fun purchaseProduct(
    activity: Activity,
    storeProduct: StoreProduct,
    purchasingOptions: StorePurchasingOptions, // Might need a better name or it could be confused with our other `PurchaseOption`
    listener: ProductChangeCallback
)
```
Where `StorePurchasingOptions` can be built following the builder pattern and contain an `UpgradeInfo`, an `isPersonalizedPrice` boolean, or both.

We would need to keep and deprecate the other older methods to avoid breaking people's apps if possible, but this seems like a more extensible option.

Reply:
you're right, it's definitely more future-proof... and, now that i think about it, amazon doesn't support upgrades/downgrades, so it'd be nice to pull that into something store-specific. i'll get to work on it!

Reply:
ok, one thing that happens with this refactor is that we don't know whether it's a product upgrade or not (you could call purchase with a `StorePurchasingOptions` and only set the `isPersonalizedPrice`), so we can't ask them to pass a  `ProductChangeCallback` instead of `PurchaseCallback`.
the difference between the two  callbacks is that `ProductChangeCallback`'s success could return a nullable Transaction (in the case of a deferred transfer).

we could just error if the wrong type of callback is passed in (i.e. a `ProductChangeCallback` when `storePurchaseOptions.upgradeInfo` is null)... 

or just omit the upgradeinfo from storePurchasingOption, though we lose some of the benefit that way..

Reply:
Ahh you’re right… We could have a new callback with a new TransacionState sealed class or something like that, where possible values are .deferred or .completed(storeTransaction). I’m afraid we are over complicating the API at that point though…

I’m leaning towards leaving it as you have it here… I think adding 2 extra entities for developers to parse for just one extra optional parameter is a lot... If we ever need to add another optional parameter we can reconsider.